File: C/1520.c
#include <stdio.h>

int main() {
	// 초기화(d는 입력받을 data, r은 출력할 result) 
	int d[172][172], r[172][172], i, j;
	int a, b, x, y, z, k;
	
	// 입력
	scanf("%d %d", &a, &b);
	scanf("%d %d %d", &x, &y, &z);
	for(i = 1; i <= a; i++)
		for(j = 1; j <= b; j++) 
		{
			scanf("%d", &d[i][j]);
			r[i][j] = 0;
		}
	scanf("%d", &k);
	
	// 연산
	while(k--) 
	{
		for(i = 1; i <= a; i++)
			for(j = 1; j <= b; j++)
			{
				// 8방향 더하기
				r[i][j] = d[i-1][j-1] + d[i-1][j] + d[i-1][j+1] + d[i][j-1] + d[i][j+1] + d[i+1][j-1] + d[i+1][j] + d[i+1][j+1];
				if(d[i][j] == 0) r[i][j] = r[i][j] == x ? 1 : 0;
				else r[i][j] = (y <= r[i][j] && r[i][j] < z);
			}
			
		// 세대교체
		for(i = 1; i <= a; i++) 
			for(j = 1; j <= b; j++)
				d[i][j] = r[i][j];
	}
	
	// 출력
	for(i = 1; i <= a; i++, puts(""))
		for(j = 1; j <= b; j++)
			printf("%d ", r[i][j]);
	
	return 0;
}

File: Java/1520.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		// 초기화(d는 입력받을 data, r은 출력할 result) 
		int d[][] = new int[172][172];
		int r[][] = new int[172][172];
		int a, b, x, y, z, k, i, j;
		
		// 입력받기
		a = sc.nextInt(); b = sc.nextInt();
		x = sc.nextInt(); y = sc.nextInt(); z = sc.nextInt();
		
		for(i = 1; i <= a; i++)
			for(j = 1; j <= b; j++) 
			{
				d[i][j] = sc.nextInt();
				r[i][j] = 0;
			}
		k = sc.nextInt();
		
		// 연산 
		while(k > 0) 
		{
			for(i = 1; i <= a; i++)
				for(j = 1; j <= b; j++)
				{
					// 8방향 더하기 
					r[i][j] = d[i-1][j-1] + d[i-1][j] + d[i-1][j+1] + d[i][j-1] + d[i][j+1] + d[i+1][j-1] + d[i+1][j] + d[i+1][j+1];
					if(d[i][j] == 0) r[i][j] = r[i][j] == x ? 1 : 0;
					else r[i][j] = (y <= r[i][j] && r[i][j] < z) ? 1 : 0;
				}
				
			// 세대 업데이트
			for(i = 1; i <= a; i++) 
				for(j = 1; j <= b; j++)
					d[i][j] = r[i][j];
			
			k--;
		}
		
		// 출력
		for(i = 1; i <= a; i++)
		{
			for(j = 1; j <= b; j++)
				System.out.printf("%d ", r[i][j]);
			System.out.println();
		}
	}
}